File: test-suite/src/test/java/io/micronaut/gcp/pubsub/subscriber/AcknowledgementPushSubscriberTest.java
package io.micronaut.gcp.pubsub.subscriber;

import io.micronaut.context.annotation.Property;
import io.micronaut.context.annotation.Requires;
import io.micronaut.context.event.BeanCreatedEvent;
import io.micronaut.context.event.BeanCreatedEventListener;
import io.micronaut.core.annotation.NonNull;
import io.micronaut.gcp.pubsub.annotation.PubSubClient;
import io.micronaut.gcp.pubsub.annotation.Topic;
import io.micronaut.gcp.pubsub.bind.DefaultPubSubAcknowledgement;
import io.micronaut.gcp.pubsub.push.PushRequest;
import io.micronaut.gcp.pubsub.support.Animal;
import io.micronaut.http.HttpRequest;
import io.micronaut.http.HttpResponse;
import io.micronaut.http.HttpStatus;
import io.micronaut.http.client.HttpClient;
import io.micronaut.http.client.annotation.Client;
import io.micronaut.http.client.exceptions.HttpClientResponseException;
import io.micronaut.json.JsonMapper;
import io.micronaut.messaging.Acknowledgement;
import io.micronaut.test.annotation.MockBean;
import io.micronaut.test.extensions.junit5.annotation.MicronautTest;
import jakarta.inject.Inject;
import jakarta.inject.Singleton;
import org.junit.jupiter.api.Assertions;
import org.junit.jupiter.api.BeforeEach;
import org.junit.jupiter.api.Test;
import org.mockito.ArgumentCaptor;
import org.mockito.Mockito;
import reactor.core.publisher.Mono;

import java.io.IOException;
import java.time.Duration;
import java.util.Base64;
import java.util.HashMap;

import static org.awaitility.Awaitility.await;
import static org.mockito.ArgumentMatchers.any;
import static org.mockito.Mockito.*;

@MicronautTest
@Property(name = "spec.name", value = "AcknowledgementPushSubscriberTest")
@Property(name = "gcp.projectId", value = "test-project")
class AcknowledgementPushSubscriberTest {

    @Inject
    @Client("/")
    HttpClient pushClient;

    @Inject
    MessageProcessor messageProcessor;

    @Inject
    AcknowledgementPushSubscriber subscriber;

    @BeforeEach
    void setup() {
        Mockito.reset(messageProcessor, subscriber);
    }

    @Test
    void testBlockingAck() throws IOException {
        when(messageProcessor.handleAnimalMessage(any())).thenReturn(Mono.just(Boolean.TRUE));

        Animal dog = new Animal("dog");
        String encodedData = Base64.getEncoder().encodeToString(JsonMapper.createDefault().writeValueAsString(dog).getBytes());
        PushRequest request = new PushRequest("projects/test-project/subscriptions/animals-push", new PushRequest.PushMessage(new HashMap<>(), encodedData, "1", "2021-02-26T19:13:55.749Z"));
        HttpResponse<?> response = pushClient.toBlocking().exchange(HttpRequest.POST("/push", request));

        Assertions.assertEquals(HttpStatus.OK, response.getStatus());
    }

    @Test
    void testBlockingNack() throws IOException {
        when(messageProcessor.handleAnimalMessage(any())).thenReturn(Mono.just(Boolean.FALSE));

        Animal cat = new Animal("cat");
        String encodedData = Base64.getEncoder().encodeToString(JsonMapper.createDefault().writeValueAsString(cat).getBytes());
        PushRequest request = new PushRequest("projects/test-project/subscriptions/animals-push", new PushRequest.PushMessage(new HashMap<>(), encodedData, "1", "2021-02-26T19:13:55.749Z"));

        try {
            pushClient.toBlocking().exchange(HttpRequest.POST("/push", request));
            Assertions.fail();
        } catch (HttpClientResponseException ex) {
            Assertions.assertEquals(HttpStatus.UNPROCESSABLE_ENTITY, ex.getResponse().getStatus());
        }
    }

    @Test
    void testAsyncAck() throws IOException {
        when(messageProcessor.handleAnimalMessage(any())).thenReturn(Mono.just(Boolean.TRUE));

        Animal dog = new Animal("dog");
        String encodedData = Base64.getEncoder().encodeToString(JsonMapper.createDefault().writeValueAsString(dog).getBytes());
        PushRequest request = new PushRequest("projects/test-project/subscriptions/animals-async-push", new PushRequest.PushMessage(new HashMap<>(), encodedData, "1", "2021-02-26T19:13:55.749Z"));
        HttpResponse<?> response = pushClient.toBlocking().exchange(HttpRequest.POST("/push", request));

        Assertions.assertEquals(HttpStatus.OK, response.getStatus());
    }

    @Test
    void testAsyncNack() throws IOException {
        when(messageProcessor.handleAnimalMessage(any())).thenReturn(Mono.just(Boolean.FALSE));

        Animal cat = new Animal("cat");
        String encodedData = Base64.getEncoder().encodeToString(JsonMapper.createDefault().writeValueAsString(cat).getBytes());
        PushRequest request = new PushRequest("projects/test-project/subscriptions/animals-async-push", new PushRequest.PushMessage(new HashMap<>(), encodedData, "1", "2021-02-26T19:13:55.749Z"));

        try {
            pushClient.toBlocking().exchange(HttpRequest.POST("/push", request));
            Assertions.fail();
        } catch (HttpClientResponseException ex) {
            Assertions.assertEquals(HttpStatus.UNPROCESSABLE_ENTITY, ex.getResponse().getStatus());
        }
    }

    @MockBean(MessageProcessor.class)
    MessageProcessor mockMessageProcessor() {
        return mock(MessageProcessor.class);
    }

    @Singleton
    @Requires(property = "spec.name", value = "AcknowledgementPushSubscriberTest")
    static class SubscriberCreatedListener implements BeanCreatedEventListener<AcknowledgementPushSubscriber> {
        @Override
        public AcknowledgementPushSubscriber onCreated(@NonNull BeanCreatedEvent<AcknowledgementPushSubscriber> event) {
            return spy(event.getBean());
        }
    }
}

File: test-suite-python/src/test/python/micronaut/gcp/pubsub/subscriber/AcknowledgementPushSubscriberTest.py
from typing import Annotated

from jakarta.inject import Inject
from java.util import Base64, HashMap
from micronaut.context.annotation import Property
from micronaut.gcp.pubsub.push import PushRequest
from micronaut.gcp.pubsub.support.Animal import Animal
from micronaut.http import HttpRequest, HttpStatus
from micronaut.http.client import HttpClient
from micronaut.http.client.annotation import Client
from micronaut.http.client.exceptions import HttpClientResponseException
from micronaut.json import JsonMapper
from micronaut.test.extensions.junit5.annotation import MicronautTest
from org.junit.jupiter.api import BeforeEach, Test

from micronaut.gcp.pubsub.subscriber.MessageProcessor import MessageProcessor


@MicronautTest
@Property(name="spec.name", value="AcknowledgementPushSubscriberTest")
@Property(name="gcp.projectId", value="test-project")
class AcknowledgementPushSubscriberTest:
    pushClient: Annotated[HttpClient, Inject, Client("/")]
    messageProcessor: Annotated[MessageProcessor, Inject]
    jsonMapper: Annotated[JsonMapper, Inject]

    @BeforeEach
    def setup(self):
        self.messageProcessor.result = True
        self.messageProcessor.received = []

    def push(self, subscription: str, animal: Animal):
        encoded_data = Base64.getEncoder().encodeToString(self.jsonMapper.writeValueAsBytes(animal))
        request = PushRequest("projects/test-project/subscriptions/" + subscription,
                              PushRequest.PushMessage(HashMap(), encoded_data, "1", "2021-02-26T19:13:55.749Z"))
        return self.pushClient.toBlocking().exchange(HttpRequest.POST("/push", request))

    @Test
    def test_blocking_ack(self):
        self.messageProcessor.result = True

        response = self.push("animals-push", Animal("dog"))

        assert response.getStatus() == HttpStatus.OK

    @Test
    def test_blocking_nack(self):
        self.messageProcessor.result = False

        try:
            self.push("animals-push", Animal("cat"))
        except HttpClientResponseException as ex:
            assert ex.getResponse().getStatus() == HttpStatus.UNPROCESSABLE_ENTITY
        else:
            assert False, "expected a 422 response"

    @Test
    def test_async_ack(self):
        self.messageProcessor.result = True

        response = self.push("animals-async-push", Animal("dog"))

        assert response.getStatus() == HttpStatus.OK

    @Test
    def test_async_nack(self):
        self.messageProcessor.result = False

        try:
            self.push("animals-async-push", Animal("cat"))
        except HttpClientResponseException as ex:
            assert ex.getResponse().getStatus() == HttpStatus.UNPROCESSABLE_ENTITY
        else:
            assert False, "expected a 422 response"
